### PR TITLE
Add batch processing to AdversarialPatch

### LINKED ISSUE #96: Unnecessary calculation of gradient of all instances in Adversarial Patch
In the `generate(self, x, y=None):` of the Adversarial Patch attack the following line calculates the gradient for the whole x vector, which contains ALL training instances:
 `gradients = self.classifier.loss_gradient(patched_images, to_categorical(np.broadcast_to(np.array(self.target), x.shape[0]),                                                                   self.classifier.nb_classes))`
However, afterwards the patch is accumulated only from 0 up to `i_batch`, it means only `gradients[0:i_batch]` is used.
Thus, why to calculate the gradients for all instances when they are not used?. I had some _Exhausted Resources Errors_ because the tensor was too big.

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

Add batch processing to AdversarialPatch

Following solution by @WladimirDelaCadena in https://github.com/IBM/adversarial-robustness-toolbox/issues/96#issuecomment-505424293

Fixes #96 

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by beat-buesser** [comment]
This PR must be merged after PR #93
